core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,253 @@
+            partitionScalingInfos[i] = new PartitionScalingInfo();
+        }
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Consumer<PageReference> wholePagePartition = partitionPage(page, partitionedPagePreparer.apply(page));
+        if (wholePagePartition != null) {
+            // whole input page will go to this partition, compact the input page avoid over-retaining memory and to
+            // match the behavior of sub-partitioned pages that copy positions out
+            page.compact();
+            sendPageToPartition(wholePagePartition, page);
+        }
+    }
+
+    private Consumer<PageReference> partitionPage(Page page, Page partitionPage)
+    {
+        // Assign each row to a partition which limits to the partitionCount. If there are more partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
+        // unique id. Thus, multiple partitions will be scaled together since we track partition physicalWrittenBytes
+        // using the artificial limit (partitionCount). The assignments lists are all expected to cleared by the
+        // previous iterations.
+        for (int position = 0; position < partitionPage.getPositionCount(); position++) {
+            int partition = partitionFunction.getPartition(partitionPage, position);
+            partitionAssignments[partition].add(position);
+        }
+
+        // Read the partition level physicalWrittenBytes. The length will be either 0 or partitionCount.
+        long[] partitionPhysicalWrittenBytes = partitionPhysicalWrittenBytesSupplier.get();
+
+        // assign each partition to a specific writer either within initial set of writers or extra unused
+        // writers if it is scaled.
+        for (int partition = 0; partition < partitionAssignments.length; partition++) {
+            IntArrayList positionsList = partitionAssignments[partition];
+            int partitionSize = positionsList.size();
+            if (partitionSize == 0) {
+                continue;
+            }
+
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            int nextPartitionWriter = getNextPartitionWriter(partition, partitionPhysicalWrittenBytes);
+
+            for (int i = 0; i < partitionSize; i++) {
+                writerAssignments[nextPartitionWriter].add(positions[i]);
+            }
+        }
+
+        // build a page for each writer
+        for (int bucket = 0; bucket < writerAssignments.length; bucket++) {
+            IntArrayList positionsList = writerAssignments[bucket];
+            int bucketSize = positionsList.size();
```

Comment:
nit: I would drop `bucketSize` and use `positionsList.size()` directly, the name is confusing.